### DISCUSSION TITLE
Add a link argument that is missing from the pkgconfig .pc file

### DISCUSSION THREAD

**Discussion by oscarbenjamin** [mesonbuild/meson-python#607] [Q&A] [answered]
This follows #602 and #604.

The remaining outstanding issue for https://github.com/flintlib/python-flint/pull/129 is that using meson-python and pkg config is incompatible with `Flint < 3.1.0`. The build appears to succeed but at runtime import fails with:
```
ImportError: /home/oscar/.pyenv/versions/python-flint-3.12/lib/python3.12/site-packages/flint/flint_base/flint_context.cpython-312-x86_64-linux-gnu.so: undefined symbol: flint_cleanup
```
Here `flint_cleanup` is just the first symbol from `libflint` that is looked up when python-flint's flint module is imported. Actually no symbols are found and so even if I comment out the use of `flint_cleanup` I just get an error with the next symbol.

It turns out that this is because of a bug in Flint's pkgconfig configuration that was fixed in Flint 3.1.0 by adding `-lflint` to `flint.pc`: https://github.com/flintlib/flint/pull/1647. I have confirmed that if I manually edit `flint.pc` to add `-lflint` then the problem is solved for Flint 3.0.1 and everything seems to work fine.

This leaves python-flint with a choice:

- Drop support for `Flint < 3.1.0`.
- Work around the buggy `flint.pc` configuration.

I'm exploring the how to work around the buggy `flint.pc` config by adding something in the meson build files to add `-lflint` manually but I have not found a solution that works. I tried adding
```
add_global_arguments('-lflint', language: 'c')
```
but that did not seem to have any effect. I also tried adding
```
link_args: '-lflint',
```
to every `extension_module` and this seemed to resolve some unrecognised symbols but not all of them and so it instead fails with:
```
ImportError: /home/oscar/.pyenv/versions/python-flint-3.12/lib/python3.12/site-packages/flint/types/fmpz.cpython-312-x86_64-linux-gnu.so: undefined symbol: partitions_fmpz_fmpz
```
I don't see exactly what is special about the function `partitions_fmpz_fmpz` in this situation:
https://github.com/flintlib/python-flint/blob/1ce152dffc356af69b1d4c2ea0eb08854f3d733b/src/flint/flintlib/partitions.pxd#L10
https://github.com/flintlib/python-flint/blob/1ce152dffc356af69b1d4c2ea0eb08854f3d733b/src/flint/types/fmpz.pyx#L709

Is there a simple way to edit the meson.build files (or otherwise) that is more or less equivalent to adding the `-lflint` argument to `flint.pc`?

**Comment by rgommers** [comment, marked as answer]
> Is there a simple way to edit the meson.build files (or otherwise) that is more or less equivalent to adding the `-lflint` argument to `flint.pc`?

Something like this using [declare_dependency](https://mesonbuild.com/Reference-manual_functions.html#declare_dependency) should work (untested, but I have code like this somewhere):
```meson
flint_dep = dependency('flint')
if flint_dep.version().version_compare('<3.1')
    flint_dep = declare_dependency(
        dependencies: flint_dep,
        link_args: ['-lflint']),
    )
endif
```

**Reply by eli-schwartz** [reply]
I would have suggested much the same except with `cc.find_library('flint')` instead of the string `-lflint`.

**Reply by oscarbenjamin** [reply]
Thanks both. Setting `-lflint` I get a link error when using the local build of flint:
```
cc  -o src/flint/pyflint.cpython-312-darwin.so src/flint/pyflint.cpython-312-darwin.so.p/meson-generated_src_flint_pyflint.pyx.c.o -Wl,-dead_strip_dylibs -Wl,-headerpad_max_install_names -Wl,-undefined,dynamic_lookup -bundle -Wl,-undefined,dynamic_lookup -Wl,-rpath,/Users/enojb/work/dev/python-flint/.local/lib /Users/enojb/work/dev/python-flint/.local/lib/libgmp.dylib /Users/enojb/work/dev/python-flint/.local/lib/libmpfr.dylib -lflint
ld: library 'flint' not found
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
Setting `LIBRARY_PATH` gets me to a working build on OSX:
```
LIBRARY_PATH=$(pwd)/.local/lib PKG_CONFIG_PATH=$(pwd)/.local/lib/pkgconfig python -m build
```
That's probably acceptable because I think it will work for someone who has a system-wide install that is already in the appropriate paths. I'll check `cc.find_library` but I guess that wold need `C_INCLUDE_DIR` etc to be set as well.

**Reply by oscarbenjamin** [reply]
> I'll check cc.find_library but I guess that wold need C_INCLUDE_DIR etc to be set as well.

No, it doesn't seem to make any difference whether I use `-lflint` or `cc.find_library`. I guess `find_library` is more robust or something so I'll go with that.

**Reply by oscarbenjamin** [reply]
Thanks both.

I have tested this and everything seems to be working in CI now for https://github.com/flintlib/python-flint/pull/129. The CI failures now are just a sporadic segfault on OSX (that I think is unrelated...) and https://github.com/mesonbuild/meson/issues/12401 which fails the 3.13-dev build.

That is enough for python-flint to have a working meson-python setup!

**Reply by eli-schwartz** [reply]
> I guess `find_library` is more robust or something so I'll go with that.

It is guaranteed to be resolved at setup time and isn't prone to changing depending on whether you return to a previously existing build directory + forget to export the same environment variables.
